private/pkg/app/appcmd/webpages.go

```diff
@@ -0,0 +1,306 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package appcmd
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"html"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"unicode"
+
+	"github.com/bufbuild/buf/private/pkg/app"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+const (
+	slugPrefixFlagName      = "slug-prefix"
+	excludeCommandsFlagName = "exclude-command"
+)
+
+var codeBlockRegex = regexp.MustCompile(`(^\s\s\s\s)|(^\t)`)
+
+type webpagesFlags struct {
+	SlugPrefix      string
+	ExcludeCommands []string
+}
+
+func newFlags() *webpagesFlags {
+	return &webpagesFlags{}
+}
+
+func (f *webpagesFlags) Bind(flagSet *pflag.FlagSet) {
+	flagSet.StringVar(
+		&f.SlugPrefix,
+		slugPrefixFlagName,
+		"",
+		"The slug prefix for front-matter slug attribute",
+	)
+	flagSet.StringSliceVar(
+		&f.ExcludeCommands,
+		excludeCommandsFlagName,
+		nil,
+		"Exclude these commands from doc generation",
+	)
+}
+
+// newWebpagesCommand returns a "webpages" command that generates docusaurus markdown for cobra commands.
+// In the future this will need to be adapted to accept a Command when cobra.Command is removed.
+func newWebpagesCommand(
+	command *cobra.Command,
+) *Command {
+	flags := newFlags()
+	return &Command{
+		Use:    "webpages",
+		Args:   cobra.ExactArgs(1),
+		Hidden: true,
+		Run: func(ctx context.Context, container app.Container) error {
+			excludes := make(map[string]bool)
+			for _, exclude := range flags.ExcludeCommands {
+				excludes[exclude] = true
+			}
+			for _, cmd := range command.Commands() {
+				if excludes[cmd.CommandPath()] {
+					cmd.Hidden = true
+				}
+			}
+			return generateMarkdownTree(
+				command,
+				container.Arg(0),
+				flags.SlugPrefix,
+			)
+		},
+		BindFlags: flags.Bind,
+	}
+}
+
+// generateMarkdownTree generates markdown for a whole command tree.
+func generateMarkdownTree(cmd *cobra.Command, dir string, slugPrefix string) error {
+	if !cmd.IsAvailableCommand() {
+		return nil
+	}
+	for _, command := range cmd.Commands() {
+		if err := generateMarkdownTree(command, dir, slugPrefix); err != nil {
+			return err
+		}
+	}
+	commandPath := commandFilePath(cmd)
+	filename := filepath.Join(dir, commandPath)
+	if err := os.MkdirAll(path.Dir(filename), os.ModePerm); err != nil {
+		return err
+	}
+	file, err := os.Create(filename)
+	if err != nil {
+		return err
+	}
+	if err := generateMarkdownPage(cmd, file, slugPrefix); err != nil {
+		return err
+	}
+	return file.Close()
```

Comment:
grep for other examples in our codebase, this is how how we handle closes. You do them in a defer with multierr.Append appending to retErr

Reply:
Yep, missed this the first time